cmd/ig/daemon.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	gadgetservice "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+func addDaemonCommand(rootCmd *cobra.Command, runtime runtime.Runtime) {
+	daemonCmd := &cobra.Command{
+		Use:   "daemon",
```

Comment:
I'm wondering about the meaning of daemon here. From [wikipedia](https://en.wikipedia.org/wiki/Daemon_(computing)) it's defined as "[computer program](https://en.wikipedia.org/wiki/Computer_program) that runs as a [background process](https://en.wikipedia.org/wiki/Background_process), rather than being under the direct control of an interactive user", but in this PR is used more like a server. Should we call it server instead of daemon? /cc @alban

Reply:
Hmm, I think in its current state "server" would indeed be a better choice, as it just serves requests and doesn't do any background running of things. With the long running gadgets that will however change again and things will indeed run in the background and you just control using what exactly that is using the socket (in addition to still be able to use it as a "server").

Reply:
What will be the daemon behaviour in the future? Will it fork and run in background? (Somehow related to Jose's comment in https://github.com/inspektor-gadget/inspektor-gadget/pull/2078#discussion_r1337571289)

Ideally the solution will be to have an `gadgetd` / `igd` and `gadgetctl` binaries, but we also want it to be able to work alone. I think the current solution is good enough, just be sure to document what daemon is about.

Reply:
xref discussion [here](https://github.com/inspektor-gadget/inspektor-gadget/pull/2078#discussion_r1337571289)

Reply:
Probably it can still be called server and the matter it's a daemon or not is controlled by systemd?, i.e. we don't provide an option to run as a daemon directly. WDYT?

/cc @alban @blanquicet pinging you folks to consolidate the discussion on this thread.

Reply:
I agree: let's call the CLI command "ig server" and let's run this in the foreground.

If ever needed later, we could add an option "ig server -d" to run in the background, but that should not be necessary with systemd, so I prefer not to add that in this PR.

Reply:
I don't have a strong opinion on this, but it seems™ like right now it's behaving the same as [`dockerd`](https://github.com/moby/moby/blob/master/cmd/dockerd/daemon.go).